### ISSUE TITLE
MQTT version 3.6.5 bug 

### ISSUE THREAD

**Issue by xue532886747** [hannesa2/paho.mqtt.android#719]
```
java.lang.RuntimeException: Error receiving broadcast Intent { act=.callbackToActivity.v0 flg=0x10 pkg=com.zg.pad (has extras) } in info.mqtt.android.service.MqttAndroidClient@44b8293
	at android.app.LoadedApk$ReceiverDispatcher$Args.lambda$getRunnable$0$android-app-LoadedApk$ReceiverDispatcher$Args(LoadedApk.java:1819)
	at android.app.LoadedApk$ReceiverDispatcher$Args$$ExternalSyntheticLambda0.run(Unknown Source:2)
	at android.os.Handler.handleCallback(Handler.java:942)
	at android.os.Handler.dispatchMessage(Handler.java:99)
	at android.os.Looper.loopOnce(Looper.java:201)
	at android.os.Looper.loop(Looper.java:288)
	at android.app.ActivityThread.main(ActivityThread.java:7937)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:569)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1019)
Caused by: java.lang.ClassCastException: org.eclipse.paho.client.mqttv3.MqttException cannot be cast to android.os.Parcelable
	at info.mqtt.android.service.MqttAndroidClient.connectionLostAction(SourceFile:14)
	at info.mqtt.android.service.MqttAndroidClient.onReceive(SourceFile:137)
	at android.app.LoadedApk$ReceiverDispatcher$Args.lambda$getRunnable$0$android-app-LoadedApk$ReceiverDispatcher$Args(LoadedApk.java:1809)
	... 9 more
```

the app is crash

**Comment by hannesa2** [comment]
There are too less info to spend time here

**Comment by xue532886747** [comment]
`package cn.chaohi.basis.mqtt;

import android.content.Context;
import android.os.Handler;
import android.os.Looper;

import com.google.gson.Gson;

import org.eclipse.paho.client.mqttv3.IMqttActionListener;
import org.eclipse.paho.client.mqttv3.IMqttAsyncClient;
import org.eclipse.paho.client.mqttv3.IMqttDeliveryToken;
import org.eclipse.paho.client.mqttv3.IMqttToken;
import org.eclipse.paho.client.mqttv3.MqttCallbackExtended;
import org.eclipse.paho.client.mqttv3.MqttConnectOptions;
import org.eclipse.paho.client.mqttv3.MqttException;
import org.eclipse.paho.client.mqttv3.MqttMessage;

import java.util.concurrent.TimeUnit;

import cn.austerlitz.common.log.LogUtils;
import info.mqtt.android.service.Ack;
import info.mqtt.android.service.MqttAndroidClient;
import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers;
import io.reactivex.rxjava3.core.Observable;
import io.reactivex.rxjava3.core.Observer;
import io.reactivex.rxjava3.disposables.Disposable;
import io.reactivex.rxjava3.schedulers.Schedulers;

/**
 * Mqtt的管理类
 */
public class MqttManager {

    private static final String TAG = "MqttManager";
    private MqttAndroidClient mqttAndroidClient;
    //MQtt账号
    private String name;
    //密码
    private String password;
    //订阅的标题
    private String topic;

    private IMessageCallBack iMessageCallBack;
    //暂时不用
    private IMqttAsyncClient iMqttAsyncClient;

    private static final MqttManager instance = new MqttManager();

    private MqttManager() {

    }

    /**
     * 设置参数
     * @param clientId   wifiMac
     * @param host       url
     * @param name       后台返的
     * @param password   后台返的
     * @param topic      主题
     */
    public void initMqttManager(Context context, String clientId, String host, String name, String password, String topic) {
        LogUtils.d("mqttManager", "clientId = " + clientId + ",host = " + host + ",name = " + name + ",password = " + password+" ,topic="+topic);
        this.name = name;
        this.password = password;
        this.topic = topic;
        this.mqttAndroidClient = new MqttAndroidClient(context, host, clientId, Ack.AUTO_ACK);
        initMqtt();
    }

    public static MqttManager getInstance() {
        return instance;
    }

    public void setDealSettleCallBack(IMessageCallBack iDealSettleCallBack) {
        iMessageCallBack = iDealSettleCallBack;
    }


    /**
     * 初始化mqtt
     */
    private void initMqtt() {
        LogUtils.d(TAG, "initMqtt()");
        mqttAndroidClient.setCallback(new MqttCallbackExtended() {
            @Override
            public void connectComplete(boolean reconnect, String serverURI) {
                LogUtils.d(TAG, "connectComplete()");
                reSubscribe();
                shutDown();
            }

            @Override
            public void connectionLost(Throwable cause) {
                LogUtils.d(TAG, "connectionLost cause = " + cause);
                if (iMessageCallBack != null) {
                    iMessageCallBack.netWorkFailure(cause);
                }
            }

            @Override
            public void messageArrived(String topic, MqttMessage message) {
                LogUtils.d(TAG, "messageArrived topic = " + topic + " , message = " + message);
                if (message != null) {
                    byte[] payload = message.getPayload();
                    String msg = new String(payload);
                    LogUtils.d(TAG, "messageArrived" + " mqttAndroidClient = " + mqttAndroidClient + " , mqttAndroidClient.isConnect = " + mqttAndroidClient.isConnected());
                    if (iMessageCallBack != null && mqttAndroidClient != null && mqttAndroidClient.isConnected()) {
                        iMessageCallBack.getMessage(msg);
                    }
                }
            }

            @Override
            public void deliveryComplete(IMqttDeliveryToken token) {
                LogUtils.d(TAG, "deliveryComplete token = " + token);

            }
        });
    }

    /**
     * 配置初始值
     */
    private MqttConnectOptions initOption() {
        MqttConnectOptions conOpt = new MqttConnectOptions();
        //清除缓存
        conOpt.setCleanSession(true);
        //自动重连
        conOpt.setAutomaticReconnect(true);
        //设置超时时间
        conOpt.setConnectionTimeout(20);
        //心跳包发送间隔
        conOpt.setKeepAliveInterval(20);
        //设置用户名
        conOpt.setUserName(name);
        //密码
        conOpt.setPassword(password.toCharArray());
        return conOpt;
    }

    /**
     * 开启mqtt的连接
     */
    public void startMqtt() {
        LogUtils.d(TAG, "startMqtt");
        try {
            if (mqttAndroidClient != null) {
                mqttAndroidClient.connect(initOption(), null, iMqttActionListener);
            }
        } catch (Exception e) {
            e.printStackTrace();
            LogUtils.e(TAG, "setMqtt mqtt异常 e1= " + e.toString());
        }
    }

    private final IMqttActionListener iMqttActionListener = new IMqttActionListener() {
        @Override
        public void onSuccess(IMqttToken asyncActionToken) {
            shutDown();
            LogUtils.d(TAG, "onSuccess: 连接成功" + "asyncActionToken = " + asyncActionToken.toString());
            try {
                //连接成功后订阅主题
                iMqttAsyncClient = asyncActionToken.getClient();
                mqttAndroidClient.subscribe(topic, 2);
            } catch (Exception e) {
                e.printStackTrace();
            }
        }

        @Override
        public void onFailure(IMqttToken asyncActionToken, Throwable exception) {
            LogUtils.d(TAG, " onFailure :连接失败 exception = " + exception);
            if (iMessageCallBack != null) {
                iMessageCallBack.msgFailure(exception);
            }
            MqttException mqttException = asyncActionToken.getException();
            if (mqttException != null) {
                int reasonCode = mqttException.getReasonCode();
                LogUtils.d(TAG, " exception reasonCode= " + reasonCode);
                if (reasonCode == MqttException.REASON_CODE_SERVER_CONNECT_ERROR) {
                    //服务器异常 ，这个时候不会再重连，应该关闭端口，重新链接
                    reconnectMqtt();
                }
            }
        }
    };

    /**
     * 重新订阅
     */
    public void reSubscribe() {
        LogUtils.d(TAG, "reSubscribe");
        if (isConnected()) {
            try {
                //连接成功后订阅主题
                if (mqttAndroidClient != null) {
                    mqttAndroidClient.subscribe(topic, 2);
                }
            } catch (Exception e) {
                e.printStackTrace();
                LogUtils.e(TAG, "setMqtt mqtt异常 e2= " + e.toString());
            }
        }
    }

    /**
     * 关闭mqtt
     */
    public synchronized void closeMqtt() {
        LogUtils.i(TAG, "closeMqtt");
        if (mqttAndroidClient != null) {
            try {
                if (iMqttAsyncClient != null && iMqttAsyncClient.isConnected()) {
                    LogUtils.e(TAG, "setMqtt iMqttAsyncClientx closeMqtt");
                    iMqttAsyncClient.close();
                    iMqttAsyncClient.disconnect();
                }
                mqttAndroidClient.disconnect();
                mqttAndroidClient.unregisterResources();
                mqttAndroidClient.close();
                shutDown();
            } catch (Exception e) {
                e.printStackTrace();
                LogUtils.e(TAG, "setMqtt mqtt异常 e3= " + e);
            }
        }
    }

    /**
     * MQTT重新连接
     */
    private void reconnectMqtt() {
        LogUtils.i(TAG, "reconnectMqtt");
        try {
            if (!isConnected()) {
                cutTimerDown();
            }
        } catch (Exception e) {
            e.printStackTrace();
            LogUtils.i(TAG, "reconnectMqtt error = " + e);
        }
    }

    /**
     * 是否链接
     *
     * @return
     */
    private boolean isConnected() {
        if (mqttAndroidClient != null && mqttAndroidClient.isConnected()) {
            return true;
        }
        return false;
    }


    public void cutTimerDown() {
        shutDown();
        LogUtils.i(TAG, "cutTimerDown开启倒计时");
        Observable.intervalRange(0, 20, 0, 1, TimeUnit.SECONDS).subscribeOn(Schedulers.computation()).observeOn(AndroidSchedulers.mainThread()).subscribe(new Observer<Long>() {
            @Override
            public void onSubscribe(@io.reactivex.rxjava3.annotations.NonNull Disposable d) {
                mDisposable = d;

            }

            @Override
            public void onNext(@io.reactivex.rxjava3.annotations.NonNull Long aLong) {

            }

            @Override
            public void onError(@io.reactivex.rxjava3.annotations.NonNull Throwable e) {

            }

            @Override
            public void onComplete() {
                if (!isConnected()) {
                    LogUtils.i(TAG, "倒计时结束，准备重连");
                    startMqtt();
                } else {
                    LogUtils.i(TAG, "倒计时结束，已连接，无需重连");
                }
            }
        });
    }

    private Disposable mDisposable;

    /**
     * 停止计时
     */
    private void shutDown() {
        LogUtils.i(TAG, "shutDown()");
        if (mDisposable != null && !mDisposable.isDisposed()) {
            mDisposable.dispose();
        }
    }

}
``package cn.chaohi.shopping.master;

import com.google.gson.Gson;

import cn.austerlitz.common.log.LogUtils;
import cn.chaohi.basis.AppData;
import cn.chaohi.basis.data.shared.LogicSharedPreferences;
import cn.chaohi.basis.mqtt.IMessageCallBack;
import cn.chaohi.basis.mqtt.MqttManager;

import cn.chaohi.shopping.BaseApp;

public class MasterManager {
    public static final String tag = "MasterManager1";


    private IDealSettleCallBack iDealSettleCallBack;

    private static final MasterManager INSTANCE = new MasterManager();

    // 提供全局访问点
    public static MasterManager getInstance() {
        return INSTANCE;
    }

    private MasterManager() {
        initView();
    }

    public void setDealSettleCallBack(IDealSettleCallBack iDealSettleCallBack) {
        this.iDealSettleCallBack = iDealSettleCallBack;
    }

    private void initView() {
        LogUtils.d(tag, "initView()");
        LogicSharedPreferences ls = new LogicSharedPreferences(BaseApp.getInstance().getApplicationContext());
        String mqttHost = ls.getMqttHost();
        String mqttName = ls.getMqttName();
        String mqttPass = ls.getMqttPass();
        String deviceId = AppData.getDeviceId();
        String ids = deviceId + "cart";
        LogUtils.d(tag, "AppData.getDeviceId() = " + ids);
        String topics = "/heart_beat/hi-cart/inspect/" + deviceId + "/status_notify";
        MqttManager.getInstance().initMqttManager(BaseApp.getInstance().getApplicationContext(), ids, mqttHost, mqttName, mqttPass, topics);
        MqttManager.getInstance().setDealSettleCallBack(new IMessageCallBack() {
            @Override
            public void getMessage(String msg) {
                try {
                    LogUtils.d(tag, "initMqtt msg = " + msg);
                    NewMasterBean newMasterBean = new Gson().fromJson(msg, NewMasterBean.class);
                    if (NewMasterTempUtil.getInstance().getMasterBean() == null || newMasterBean == null) {
                        LogUtils.d(tag, "对象为null");
                        return;
                    }
                    if (!NewMasterTempUtil.getInstance().getMasterBean().getInspectOrderNo().equals(newMasterBean.getInspectOrderNo())) {
                        LogUtils.d(tag, "订单号不同");
                        return;
                    }
                    if (iDealSettleCallBack != null) {
                        if (newMasterBean.getStatus() == 1 || newMasterBean.getStatus() == 2) {
                            LogUtils.d(tag, "管理员正在进行流程 status = " + newMasterBean.getStatus());
                            iDealSettleCallBack.changeStatus(newMasterBean);
                        } else if (newMasterBean.getStatus() == 3) {
                            LogUtils.d(tag, "校验通过，准备关闭弹窗");
                            iDealSettleCallBack.adminPass(newMasterBean);
                        }
                    }
                } catch (Exception e) {
                    e.printStackTrace();
                    LogUtils.d(tag, "initMqtt e = " + e.toString());
                }
            }

            @Override
            public void msgFailure(Throwable exception) {
                LogUtils.d(tag, "msgFailure exception = " + exception);
            }

            @Override
            public void netWorkFailure(Throwable exception) {
                //断网
                if (iDealSettleCallBack != null) {
                    iDealSettleCallBack.taskFailure();
                }
            }
        });
    }

    public void start() {
        MqttManager.getInstance().startMqtt();
    }

    public void clear() {
        LogUtils.d(tag, "clear master");
        MqttManager.getInstance().closeMqtt();
        NewMasterTempUtil.getInstance().setMasterBean(null);
        if (iDealSettleCallBack != null) {
            iDealSettleCallBack = null;
        }
    }
}
`When I initialize a MasterManager object in the onCreate method of an activity and the Matt link is successful, if the network is disconnected at this time, it will cause a crash，But when I use version 3.6.4, it won't crash

**Comment by hannesa2** [comment]
Thank you, this is a step in the right direction, but I mean a **running** example. You have to minimize my afford to get help and I don't assemble a reproduceable sample. Even you find this too annoying, and simply throw a non running code sample to me. Doing this, the afford is on my side.

A **running** sample means: Either a pull request with a sample in this repository or a repository where I can clone, build, run and see issue.

**Comment by dprilutskii** [comment]
@hannesa2 Hello! I encountered the same problem, and it occurs if the MQTT broker becomes unavailable for some reason. For example, a network disconnection or the broker being stopped. It is easy to reproduce in any test environment. If this information is not sufficient, I will provide you with the code.

`19:13:43.761  E  FATAL EXCEPTION: main
                 Process: my.test.app.app, PID: 6786
                 java.lang.RuntimeException: Error receiving broadcast Intent { act=.callbackToActivity.v0 flg=0x10 pkg=my.test.app.app (has extras) } in info.mqtt.android.service.MqttAndroidClient@fc402d1
                 	at android.app.LoadedApk$ReceiverDispatcher$Args.lambda$getRunnable$0$android-app-LoadedApk$ReceiverDispatcher$Args(LoadedApk.java:1822)
                 	at android.app.LoadedApk$ReceiverDispatcher$Args$$ExternalSyntheticLambda0.run(Unknown Source:2)
                 	at android.os.Handler.handleCallback(Handler.java:942)
                 	at android.os.Handler.dispatchMessage(Handler.java:99)
                 	at android.os.Looper.loopOnce(Looper.java:201)
                 	at android.os.Looper.loop(Looper.java:288)
                 	at android.app.ActivityThread.main(ActivityThread.java:7962)
                 	at java.lang.reflect.Method.invoke(Native Method)
                 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:550)
                 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:952)
                 Caused by: java.lang.ClassCastException: org.eclipse.paho.client.mqttv3.MqttException cannot be cast to android.os.Parcelable
                 	at info.mqtt.android.service.MqttAndroidClient.connectionLostAction(MqttAndroidClient.kt:1304)
                 	at info.mqtt.android.service.MqttAndroidClient.onReceive(MqttAndroidClient.kt:905)
                 	at android.app.LoadedApk$ReceiverDispatcher$Args.lambda$getRunnable$0$android-app-LoadedApk$ReceiverDispatcher$Args(LoadedApk.java:1812)
                 	at android.app.LoadedApk$ReceiverDispatcher$Args$$ExternalSyntheticLambda0.run(Unknown Source:2) 
                 	at android.os.Handler.handleCallback(Handler.java:942) 
                 	at android.os.Handler.dispatchMessage(Handler.java:99) 
                 	at android.os.Looper.loopOnce(Looper.java:201) 
                 	at android.os.Looper.loop(Looper.java:288) 
                 	at android.app.ActivityThread.main(ActivityThread.java:7962) 
                 	at java.lang.reflect.Method.invoke(Native Method) 
                 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:550) 
                 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:952) `

**Comment by hannesa2** [comment]
Version 3.6.5 comes without line 1304 😞
`info.mqtt.android.service.MqttAndroidClient.connectionLostAction(MqttAndroidClient.kt:1304)`
But I guess I found something. Legacy branch v3.x misses one commit https://github.com/hannesa2/paho.mqtt.android/pull/725

**Comment by hannesa2** [comment]
Here you go https://github.com/hannesa2/paho.mqtt.android/releases/tag/3.6.6

**Comment by dprilutskii** [comment]
@hannesa2 Yes, the bug does not occur in the beta version. We are eagerly awaiting the release version.

I would like to discuss an issue with using DisconnectedBufferOptions. This set of options is applicable only after a successful connection, and if you attempt to assign the set of options before calling the connect function, an NPE occurs (which makes sense based on the code). In the beta (4.*) version, this logic is repeated. It seems that this set of options should be applied before the connect function and should allow saving client messages until a successful connection to the broker is established. I can either initiate a new issue or try to make changes to your code.

For example:
``` java


MqttAndroidClient client = new MqttAndroidClient(this, brokerUri, CLIENT_ID, Ack.AUTO_ACK, new MqttDefaultFilePersistence(userDirectory), false, 1000);

DisconnectedBufferOptions bufferOptions = new DisconnectedBufferOptions();
            bufferOptions.setBufferEnabled(true);
            bufferOptions.setPersistBuffer(true);
            bufferOptions.setBufferSize(10_000);
            bufferOptions.setDeleteOldestMessages(false);

client.setBufferOpts(bufferOptions);

MqttConnectOptions options = new MqttConnectOptions();
options.setCleanSession(false);
options.setAutomaticReconnect(true);
options.setUserName(user);
options.setPassword(password.toCharArray());

client.client.connect(options, null, new IMqttActionListener() {
...

```

**Comment by hannesa2** [comment]
> We are eagerly awaiting the release version.

It's just number cosmetic. I removed the `*beta*`

**Comment by hannesa2** [comment]
> It seems that this set of options should be applied before the connect function and should allow saving client messages until a successful connection to the broker is established. I can either initiate a new issue or try to make changes to your code.

Feel free to make a pull request
@dprilutskii
